jobs/cloud_replicate.Jenkinsfile

```diff
@@ -0,0 +1,147 @@
+def pipeutils, pipecfg, libcloud
+node {
+    checkout scm
+    pipeutils = load("utils.groovy")
+    pipecfg = pipeutils.load_pipecfg()
+    libcloud = load("libcloud.groovy")
+}
+
+properties([
+    pipelineTriggers([]),
+    parameters([
+      choice(name: 'STREAM',
+             choices: pipeutils.get_streams_choices(pipecfg),
+             description: 'CoreOS stream to release'),
+      string(name: 'VERSION',
+             description: 'CoreOS version to release',
+             defaultValue: '',
+             trim: true),
+      string(name: 'ADDITIONAL_ARCHES',
+             description: "Override additional architectures (space-separated). " +
+                          "Supported: ${pipeutils.get_supported_additional_arches().join(' ')}",
+             defaultValue: "",
+             trim: true),
+      booleanParam(name: 'ALLOW_MISSING_ARCHES',
+                   defaultValue: false,
+                   description: 'Allow release to continue even with missing architectures'),
+      string(name: 'COREOS_ASSEMBLER_IMAGE',
+             description: 'Override coreos-assembler image to use',
+             defaultValue: "",
+             trim: true)
+    ] + pipeutils.add_hotfix_parameters_if_supported()),
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+])
+
+def build_description = "[${params.STREAM}]"
+
+// Reload pipecfg if a hotfix build was provided. The reason we do this here
+// instead of loading the right one upfront is so that we don't modify the
+// parameter definitions above and their default values.
+if (params.PIPECFG_HOTFIX_REPO || params.PIPECFG_HOTFIX_REF) {
+    node {
+        pipecfg = pipeutils.load_pipecfg(params.PIPECFG_HOTFIX_REPO, params.PIPECFG_HOTFIX_REF)
+        build_description = "[${params.STREAM}-${pipecfg.hotfix.name}]"
+    }
+}
+
+// no way to make a parameter required directly so manually check
+// https://issues.jenkins-ci.org/browse/JENKINS-3509
+if (params.VERSION == "") {
+    throw new Exception("Missing VERSION parameter!")
+}
+
+// runtime parameter always wins
+def cosa_img = params.COREOS_ASSEMBLER_IMAGE
+cosa_img = cosa_img ?: pipeutils.get_cosa_img(pipecfg, params.STREAM)
+def basearches = params.ADDITIONAL_ARCHES.split() as List
+basearches = basearches ?: pipecfg.additional_arches
+
+// we always release for x86_64
+basearches += 'x86_64'
+// make sure there are no duplicates
+basearches = basearches.unique()
+
+def stream_info = pipecfg.streams[params.STREAM]
+
+build_description += "[${basearches.join(' ')}][${params.VERSION}]"
+currentBuild.description = "${build_description} Waiting"
+
+// We just lock here out of an abundance of caution in case somehow
+// two jobs run for the same stream, but that really shouldn't happen.
+// Also lock version-arch-specific locks to make sure these builds are finished.
+lock(resource: "cloud-replicate-${params.STREAM}") {
```

Comment:
I don't think we need this lock.

Reply:
you're probably right, but we definitely wouldn't want two jobs with same stream/version to run at the same time (on accident), right?

Reply:
Same version, no. Same stream... I guess it's uncommon but perfectly safe, so I wouldn't go out of my way to block it. E.g. maybe you've already started a replication job for a version but then another build comes through that supersedes it but the replicate job isn't done yet and it's too late to cancel it without leaking resources.

I guess the simplest patch is to do `s/STREAM/VERSION/`.

Reply:
done